stl/inc/numeric

```diff
@@ -70,19 +70,30 @@ inline constexpr bool _Plus_on_arithmetic_ranges_reduction_v = false;
 #endif // _STD_VECTORIZE_WITH_FLOAT_CONTROL
 
 template <class _InIt, class _Ty, class _BinOp>
-_NODISCARD _Ty reduce(const _InIt _First, const _InIt _Last, _Ty _Val, _BinOp _Reduce_op) {

```

Comment:
Thisis really messy, so I would suggest to duplicate the function signature. Also note that `__cpp_lib_is_constant_evaluated` requires `_HAS_CXX20` so there is no need to duplicated that here

```cpp
template <class _InIt, class _Ty, class _BinOp>
#ifdef __cpp_lib_is_constant_evaluated
_NODISCARD _CONSTEXPR20 _Ty reduce(const _InIt _First, const _InIt _Last, _Ty _Val, _BinOp _Reduce_op) {
#else // ^^^ __cpp_lib_is_constant_evaluated / !__cpp_lib_is_constant_evaluated vvv
_NODISCARD _Ty reduce(const _InIt _First, const _InIt _Last, _Ty _Val, _BinOp _Reduce_op) {
#endif // __cpp_lib_is_constant_evaluated
```

Reply:
To be honest, I agree that it looks dirty, and I'm not sure how to do it better.

Maybe?
```cpp
template <class _InIt, class _Ty, class _BinOp>
#ifdef __cpp_lib_is_constant_evaluated
_NODISCARD _CONSTEXPR20 _Ty reduce(
#else // ^^^ __cpp_lib_is_constant_evaluated / !__cpp_lib_is_constant_evaluated vvv
_NODISCARD _Ty reduce(
#endif // __cpp_lib_is_constant_evaluated
    const _InIt _First, const _InIt _Last, _Ty _Val, _BinOp _Reduce_op) {
```

```cpp
template <class _InIt1, class _InIt2, class _Ty, class _BinOp1, class _BinOp2>
#ifdef __cpp_lib_is_constant_evaluated
_NODISCARD _CONSTEXPR20 _Ty transform_reduce(
#else // ^^^ __cpp_lib_is_constant_evaluated / !__cpp_lib_is_constant_evaluated vvv
_NODISCARD _Ty transform_reduce(
#endif // __cpp_lib_is_constant_evaluated
    _InIt1 _First1, _InIt1 _Last1, _InIt2 _First2, _Ty _Val, _BinOp1 _Reduce_op, _BinOp2 _Transform_op) {
```

Reply:
Same as https://github.com/microsoft/STL/issues/395
```cpp
template <class _InIt, class _Ty, class _BinOp>
_NODISCARD _CONSTEXPR20 _Ty reduce(const _InIt _First, const _InIt _Last, _Ty _Val, _BinOp _Reduce_op) {
```

Reply:
I suggest something like this:

Next to where _CONSTEXPR20 is declared:

```
#ifdef __cpp_lib_is_constant_evaluated
#define _CONSTEXPR_ICE constexpr
#else // ^^^ __cpp_lib_is_constant_evaluated // !__cpp_lib_is_constant_evaluated vvv
#define _CONSTEXPR_ICE
#endif // __cpp_lib_is_constant_evaluated
```

Reply:
Given that this will come up more often this seems indeed a great idea :+1:

Reply:
Define in `yvals_core.h `
```
// Functions that became constexpr in C++20, and requires is_constant_evaluated
#ifdef __cpp_lib_is_constant_evaluated
#define _CONSTEXPR20_ICE constexpr
#else // ^^^ constexpr in C++20 witn is_constant_evaluated / inline (not constexpr) in C++20 and earlier without is_constant_evaluated vvv
#define _CONSTEXPR20_ICE inline
#endif // ^^^ inline (not constexpr) in C++20 and earlier without is_constant_evaluated ^^^
```